cylc/uiserver/resolvers.py

```diff
@@ -15,10 +15,189 @@
 
 """GraphQL resolvers for use in data accessing and mutation of workflows."""
 
+import getpass
+import logging
+import os
+from shlex import quote
+from subprocess import Popen, PIPE, DEVNULL
+
 from cylc.flow.network.resolvers import BaseResolvers
 from cylc.flow.data_store_mgr import WORKFLOW
 
 
+DEBUG = True
+logger = logging.getLogger(__name__)
+
+
+def snake_to_kebab(snake):
+    """Convert snake_case text to --kebab-case text.
+
+    Examples:
+        >>> snake_to_kebab('foo_bar_baz')
+        '--foo-bar-baz'
+        >>> snake_to_kebab('')
+        ''
+        >>> snake_to_kebab(None)
+        Traceback (most recent call last):
+        TypeError: <class 'NoneType'>
+
+    """
+    if isinstance(snake, str):
+        if not snake:
+            return ''
+        return f'--{snake.replace("_", "-")}'
+    raise TypeError(type(snake))
+
+
+def check_cylc_version(version):
+    """Check the provided Cylc version is available on the CLI.
+
+    Sets CYLC_VERSION=version and tests the result of cylc --version
+    to make sure the requested version is installed and selectable via
+    the CYLC_VERSION environment variable.
+    """
+    proc = Popen(
+        ['cylc', '--version', '--color=never'],
+        env={**os.environ, 'CYLC_VERSION': version},
+        stdin=DEVNULL,
+        stdout=PIPE,
+        stderr=PIPE,
+        text=True
+    )
+    ret = proc.wait(timeout=5)
+    out, err = proc.communicate()
+    return ret or out.strip() == version
+
+
+class Services:
+    """Services provided by the UI Server."""
```

Comment:
When I first read it I thought these were [JupyterHub services](https://jupyterhub.readthedocs.io/en/stable/reference/services.html) :rofl:

Reply:
Good point, will clarify.

We called these "services" back at the beginning of the Cylc8 program which is why I named it this here, however, they aren't really services any more, just methods. Perhaps we need a better name.